streams/src/test/java/org/apache/kafka/streams/integration/IQv2VersionedStoreIntegrationTest.java

```diff
@@ -153,14 +192,189 @@ private void shouldHandleVersionedKeyQuery(final Integer key,
+            try (final VersionedRecordIterator<Integer> iterator = partitionResultsEntry.getValue().getResult()) {
+                int i = LAST_INDEX;
+                int iteratorSize = 0;
+
+                // step 1:
+                while (iterator.hasNext()) {
+                    final VersionedRecord<Integer> record = iterator.next();
+                    final Long timestamp = record.timestamp();
+                    final Optional<Long> validTo = record.validTo();
+                    final Integer value = record.value();
+
+                    final Optional<Long> expectedValidTo = i < LAST_INDEX ? Optional.of(RECORD_TIMESTAMPS[i + 1]) : Optional.empty();
+                    assertThat(value, is(RECORD_VALUES[i]));
+                    assertThat(timestamp, is(RECORD_TIMESTAMPS[i]));
+                    assertThat(validTo, is(expectedValidTo));
+                    assertThat(queryResult.getExecutionInfo(), is(empty()));
+                    i--;
+                    iteratorSize++;
+                    if (i == 2)
+                        break;
+                }
+
+                // update the value of the oldest record
+                updateRecordValue();
+
+                // step 2: continue reading records from through the already opened iterator
+                while (iterator.hasNext()) {
+                    final VersionedRecord<Integer> record = iterator.next();
+                    final Long timestamp = record.timestamp();
+                    final Optional<Long> validTo = record.validTo();
+                    final Integer value = record.value();
+
+                    final Optional<Long> expectedValidTo = i < LAST_INDEX ? Optional.of(RECORD_TIMESTAMPS[i + 1]) : Optional.empty();
+                    assertThat(value, is(RECORD_VALUES[i]));
+                    assertThat(timestamp, is(RECORD_TIMESTAMPS[i]));
+                    assertThat(validTo, is(expectedValidTo));
+                    assertThat(queryResult.getExecutionInfo(), is(empty()));
+                    i--;
+                    iteratorSize++;
+                }
+
+                // The number of returned records by query is equal to expected number of records
+                assertThat(iteratorSize, equalTo(RECORD_NUMBER));
+            }
+        }
+    }
+
+    private void updateRecordValue() {
+        // update the record value at RECORD_TIMESTAMPS[0]
+        final Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+        try (final KafkaProducer<Integer, Integer> producer = new KafkaProducer<>(producerProps)) {
+            producer.send(new ProducerRecord<>(INPUT_TOPIC_NAME, 0, RECORD_TIMESTAMPS[0], RECORD_KEY, 999999));
```

Comment:
> Why do we re-use `RECORD_TIMESTAMPS[0]` here? Not sure if I can follow? -- Won't this overwrite the first/oldest value which was already returned in the first `while` loop before `updateRecordValue` is executed?

Yes, it does (it overwrites the oldest value). That 's why, when not using snapshots, the test will fail because it expects the new value (`999999` instead of the old value `RECORD_TVALUES[0])

Reply:
> which was already returned in the first `while` loop before `updateRecordValue` is executed?

No, the order is ANY. It means default, which is DESCENDING. Therefore, the first while loop just iterate over the values of the `latestValueStore`.

Reply:
Ah. Thanks. Might be worth to add a comment.